storage/filesystem/object.go

```diff
@@ -545,6 +545,9 @@ func (s *ObjectStorage) HashesWithPrefix(prefix []byte) ([]plumbing.Hash, error)
 
 	// TODO: This could be faster with some idxfile changes,
 	// or diving into the packfile.
+	if err := s.requireIndex(); err != nil {
+		return nil, err
+	}
```

Comment:
For the failing test, before `s.requireIndex()` is executed the hash is already in `hashes`. Now with `requireIndex()` the follow up code appends the same hash again, hence the duplication. The original version of `TestHashesWithPrefix` was correct.

Please add a short-circuit ahead of the `requireIndex` as that fixes the problem and yield faster results in some execution paths:

```suggestion
	// short-circuit if hashes were already found.
	if len(hashes) > 0 {
		return hashes, nil
	}

	if err := s.requireIndex(); err != nil {
		return nil, err
	}
```

This is based on the `TestHashesWithPrefix` example:
```
BenchmarkHashesWithPrefix_ReturnIfFound-16         	   15084	     78833 ns/op	   67402 B/op	    2090 allocs/op
BenchmarkHashesWithPrefix_RequireIndexOnly-16    	    1092	   1072770 ns/op	  198198 B/op	    2941 allocs/op
```

Reply:
Good questions. From what I understood from the code, in the `s.index` we have the full index - which is being reloaded by your changes. The former, I am not completely sure as in some manual tests I could confirm it does not contain some of the references, and therefore would not return all refs.

Due to that, I think you should remove:

```
	hashes, err := s.dir.ObjectsWithPrefix(prefix)
	if err != nil {
		return nil, err
	}
```

And replace it with the creation of the `hashes` array instead. IMO, the potential performance hit is overshadowed by the accuracy improvement on that case.

Reply:
That would therefore remove the need for the short-circuit as well.

Reply:
Looking at `EncodedObject(...)` It looks like its possible that there are objects that have not been packed, I think the only way to guarantee it returns all the possible values is to search both and deduplicate the results. Are you happy for me to implement that?

Reply:
Similar story with `IterEncodedObjects`, it first iterates the encoded objects, and then iterates the pack files, but passes in a list of objects that it has already `seen`.

Reply:
I've pushed an update where it is checking both, but removes duplicates the same way the rest of the code does.

Reply:
Looks good to me. Do you mind writing some comments to that effect please?

Reply:
Done, followed the comments on the other functions